File: py/fifo.py
"""Python version of FIFO implementation. It is a preview of what the C code should look like."""
# Design parameters
DEPTH = 8
WLEN = 32
# Derived constants
BOUND_LO = -2 ** (WLEN - 1)
BOUND_HI = 2 ** (WLEN - 1) - 1


class Fifo:
    """FIFO class. Slotted class, for gaining performance and fidelity of implementation."""
    __slots__ = "_mem", "_addr_rd", "_addr_wr", "is_empty"

    # ------------- #
    # BASIC METHODS #
    # ------------- #

    def __init__(self):
        self._mem = [0 for _ in range(DEPTH)]
        self._addr_rd = 0
        self._addr_wr = 0
        self.is_empty = True

    def write(self, val):
        """Insert a new value. Raise an error if not possible."""
        self._assert_wr(val)       # Check input value and available space
        self._mem[self._addr_wr] = int(val)
        self._addr_wr += 1
        self._addr_wr %= DEPTH
        self.is_empty = False

    def read(self) -> int:
        """Get the oldest entry in the queue. Raise an error if empty."""
        if self.is_empty:
            raise BufferError("FIFO is empty.")
        val = self._mem[self._addr_rd]
        self._addr_rd += 1
        self._addr_rd %= DEPTH
        if self._addr_rd == self._addr_wr:
            self.is_empty = True
        return val

    # ---------- #
    # PROPERTIES #
    # ---------- #

    @property
    def queue_len(self):
        """Get the number of elements queued."""
        q_len = (self._addr_wr - self._addr_rd) % DEPTH
        if q_len == 0:
            return 0 if self.is_empty else DEPTH
        return q_len

    @property
    def queue(self) -> int:
        """Get FIFO's queue, oldest-to-newest."""
        mem_cp = self._mem[self._addr_rd:] + self._mem[:self._addr_rd]
        return tuple(mem_cp[:self.queue_len])

    @property
    def is_full(self) -> bool:
        """Check if FIFO is full."""
        return not self.is_empty and (self._addr_rd == self._addr_wr)

    # --------------- #
    # PRIVATE METHODS #
    # --------------- #

    def _assert_wr(self, val):
        """Raise an error if the value to be written is invalid."""
        if not (BOUND_LO <= val <= BOUND_HI):
            raise ValueError(f"Value inserted out of bounds: |{val}| > {2**WLEN -1}")
        if val % 1:
            raise ValueError(f"Value inserted is not an integer: {val}")
        if self.is_full:
            raise BufferError(f"FIFO is full. ({DEPTH} words)")

if __name__ == "__main__":
    print("Python version of FIFO")

File: c/fifo.c
#include <stdio.h>
#include <string.h>
#include "fifo.h"

// TO READ: "Using GCC for compilation" - https://docs.openeuler.org/en/docs/20.03_LTS/docs/ApplicationDev/using-gcc-for-compilation.html#basics

// Get current length of fifo
int get_queue_len(fifo_t* fifo) {
    switch (fifo->status) {
        case EMPTY:
            return 0;
        case FULL:
            return DEPTH;
        case PARTIAL:
        default:
            return (fifo->addr_wr - fifo->addr_rd) % DEPTH;
    }
}

// A pseudo-constructor for Fifo
void init_fifo(fifo_t* fifo) {
    fifo->addr_rd = 0;
    fifo->addr_wr = 0;
    fifo->status = EMPTY;
}

// Insert value into fifo
signed char write(fifo_t* fifo, long val) {
    // If full, flag as an error
    if (fifo->status == FULL) {
        return OVERFLOW;
    }
    fifo->internal_mem[fifo->addr_wr] = val;
    fifo->addr_wr = (fifo->addr_wr + 1) % DEPTH;
    fifo->status = (fifo->addr_wr == fifo->addr_rd) ? FULL : PARTIAL;
    return fifo->status;
}

// Retrieve value from fifo
signed char read(fifo_t* fifo, long* val){
    if (fifo->status == EMPTY) {
        return UNDERFLOW;
    }
    *val = fifo->internal_mem[fifo->addr_rd];
    fifo->addr_rd = (fifo->addr_rd + 1) % DEPTH;
    fifo->status = (fifo->addr_rd == fifo->addr_wr) ? EMPTY : PARTIAL;
    return fifo->status;
}

// Sanity check
int main() {

#ifdef __STDC_VERSION__
    printf("C standard version: %lu\n",__STDC_VERSION__);
#endif

    long val;
    fifo_t fifo;
    init_fifo(&fifo);
    printf("Created FIFO with depth %d\n", DEPTH);
    write(&fifo, 42);
    printf("Current queue length: %d\n", get_queue_len(&fifo));
    read(&fifo, &val);
    printf("Read value: %ld\n", val);
    return 0;
}
